miner/miner.go

```diff
@@ -192,3 +204,138 @@ func (miner *Miner) SetEtherbase(addr common.Address) {
 func (self *Miner) SubscribePendingLogs(ch chan<- []*types.Log) event.Subscription {
 	return self.worker.pendingLogsFeed.Subscribe(ch)
 }
+
+// commitmentCacheSaved will check to see if the last commitment cache entry is saved.
+// If not, it will search for it and save it.
+func (miner *Miner) commitmentCacheSaved(istEngine consensus.Istanbul) bool {
+	// Subscribe to new block notifications
+	newBlockCh := make(chan core.ChainEvent)
+	bc := miner.eth.BlockChain()
+	newBlockSub := bc.SubscribeChainEvent(newBlockCh)
+	defer newBlockSub.Unsubscribe()
+
+	// getCurrentBlockAndState
+	currentBlock := bc.CurrentBlock()
+	currentHeader := currentBlock.Header()
+	currentState, err := bc.StateAt(currentBlock.Root())
+	if err != nil {
+		log.Error("Error in retrieving state", "block hash", currentHeader.Hash(), "error", err)
+		return false
+	}
+
+	// Check to see if we already have the commitment cache
+	lastCommitment, err := random.GetLastCommitment(miner.coinbase, currentHeader, currentState)
+	if err != nil {
+		log.Error("Error in retrieving last commitment", "error", err)
+		return false
+	}
+
+	// If there was no previous commitment from this validator, then return true.
+	if (lastCommitment == common.Hash{}) {
+		return true
+	}
+
+	if (rawdb.ReadRandomCommitmentCache(miner.db, lastCommitment) != common.Hash{}) {
+		return true
+	}
+
+	// goroutine communication channels and waitgroup
+	stopCh := make(chan struct{})
+	errCh := make(chan error)
+	foundParentHashCh := make(chan common.Hash)
+	var wg sync.WaitGroup
+
+	// Stacked defers are exececuted in LIFO order
+	defer wg.Wait()
+	defer close(stopCh)
+
+	if currentHeader.Number.Uint64() > 0 {
+		blockHashIter := currentHeader.Hash()
+
+		wg.Add(1)
+		go func() {
```

Comment:
i would extract this two function into their own functions: `getMostRecentBlockFromAuthor(errgroup)` and `waitForNextBlockFromAuthor(errgroup)`

and use an errgroup to handle coordination (https://godoc.org/golang.org/x/sync/errgroup) 

Also, try move evertyhing to the `random.go` file; so to confine all code related to randomness to a single place. We can even not make it a part of `miner` object, and just pass everything as parameter to a static function

Reply:
I actually was considering putting some of this logic within `istanbul`, but decided against it.  The reason is that I believe Istanbul should only be responsible for defining the randomness beacon protocol (e.g. when/how is the randomness generated and verified). 

As for the randomness cache, the only users of it is the `miner` package, as the cache is not an essential part of the randomness protocol, but more of an optimization that the `miner` can use.  So with that in mind, restoration of the cache seems to fit within the `miner` package.  I'm happy to create a new `random.go` file within `miner` package, though.